docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,317 @@
+
+**Stage 2 - minimize retypecheck in slightly modified projects**
+
+That is, when rebuilding the project after modifying an independent file, only the modified file will be retypechecked, for others the check will be omitted. This will handle many target scenarios for the feature, like modifying one test in a test project.
+
+**Stage 3 - omit reresolving assemblies**
+
+That is, for either of the scenarios above, skip the `Import mscorlib+FSharp.Core` step. This logic is also done in `main1`, prior to parsing. Same ideas apply - if cached, deserialize the information - otherwise, follow the normal process and then cache the results. The data should be stored separately from the typecheck info, so that the information be reused partially.
+
+**Stage 4 - omit reoptimization**
+
+That is, for everything that doesn't need to be retypechecked, don't do reoptimization either. This happens in the `main3` phase of compilation. 
+
+### What else can be done
+
+**Omit rewriting .NET binaries**
+
+This means that if _for some reason_ the recompilation is triggered but there is nothing to retypecheck, reuse the binaries. This shouldn't be hard to do (just writing and reading existing files) - but it's unclear how often this happens in real life.
+
+**Omit all the other steps**
+
+If we have universal caching mechanism (see below), we can apply it to other steps (e.g. parsing). Those steps typically don't take much time each on their own, but can somewhat accummulate and maybe even peak in some edge cases. 
+
+**Integrate this into tooling**
+
+We can plug the feature into the tooling. This would lead to time savings in many more developer scenarios. However, this is not trivial and the tooling layer of the feature would need to do extra checks and invalidation calls, e.g. when the tooling and the compiler use different versions of the compiler service.
+
+**Omit recompilation for insignificant file changes**
+
+That is about getting smart after getting first parts of the compilation results. If we know that the code is IL-same, we can skip everything and reuse the binaries. This would be useful in scenarios like adding comments or removing unused opens. However, this might present a **hard** (albeit interesting) challenge - things like figuring out AST diffs seem to be unexplored yet.
+
+## Open questions
+
+### How to cache data?
+
+We can store the data in the intermediate build directory (e.g. `obj` in case of `dotnet build`). The usual way for caching would be to serialize + deserialize information.
```

Comment:
How?
What and how many files will be in the directory?

(compiler flags, dependency graph, data from system imports...)

Reply:
E.g. a sample project:

A.fsi
A.fs
B.fs